### PR TITLE
fix: in a string with '&' and '!' together isn't a comment

### PR DESCRIPTION
**Pull request by gxyd**
## Description

Fixes https://github.com/lfortran/lfortran/issues/4770

### CODE REVIEW

**Comment by gxyd** [comment]
Actually, I figured another test case, which even this PR doesn't fix:
```f90
textblock = "apple'&!apple"
```

I'm looking at some other issue, but I'll have to come back to why that fails (probably because of `in_string`), happy to receive any help on this PR.

**Comment by kmr-srbh** [comment]
You are right, this happens due to `in_string` only, when we set it here:

```c++
if (s[pos] == '\'' || s[pos] == '"') {
  in_string = !in_string;
}
```

I think we need to remember the correct opening quotes. I am able to pass this using the below hack. It is not clean.

```diff
diff --git a/src/lfortran/parser/parser.cpp b/src/lfortran/parser/parser.cpp
index 6b57bfd6d..efa201a4d 100644
--- a/src/lfortran/parser/parser.cpp
+++ b/src/lfortran/parser/parser.cpp
@@ -246,6 +246,7 @@ void cont1(
 ) {
     ws_or_comment = true;
     bool in_comment = false;
+
     while (s[pos] != '\n') {
         // in a string if '&!' appear together then it isn't a comment
         if (s[pos] == '!' && (!in_string || s[pos - 1] != '&')) {
@@ -627,9 +628,16 @@ std::string prescan(const std::string &s, LocationManager &lm,
         std::string out;
         size_t pos = 0;
         bool in_comment = false, newline = true, in_string = false;
+        char quote = '\0';
         while (pos < s.size()) {
-            if (s[pos] == '\'' || s[pos] == '"') {
-                in_string = !in_string;
+            if ((s[pos] == '\'' || s[pos] == '"')) {
+                if (quote == '\0') {                // string starts, store it's opening quote
+                    in_string = true;
+                    quote = s[pos];
+                } else if (quote == s[pos]) {       // string ends, reset quote
+                    in_string = false;
+                    quote = '\0';
+                }
             }
             if (newline && is_include(s, pos)) {
                 int col = 0; // doesn't matter
```

**Comment by gxyd** [comment]
I also tried a solution to the above (I haven't pushed a solution yet) and I also tried yours as well, and then the below fails:
```f90
    textblock = "apple'&!apple"
    if (textblock /= "apple'&!apple") error stop   ! this is where it fails
```

**Comment by gxyd** [comment]
Minimal example:
```f90
program line_continuation_04
    implicit none
    character(len=20) :: textblock
    textblock = "apple&
!apple
"
    print *, textblock

    textblock = "apple'&!apple"
    if (textblock /= "apple'&!apple") error stop
end program line_continuation_04

```

**Comment by certik** [comment]
Those are some real corner cases, didn't even know about this. Just fix them one by one, add a test, etc.

**Comment by certik** [comment]
CI has to pass.

**Comment by gxyd** [comment]
> Those are some real corner cases, didn't even know about this. Just fix them one by one, add a test, etc.

Makes sense, I'll open a separate issue for once I get the PR in good shape and if that allows us to move forward with compilation of `fman` third party repository.

**Comment by gxyd** [comment]
I've mostly focused on free form prescanning in this PR, as these are quite a few corner cases, probably I'll open separate issue for free form as well, though I'll not work on it currently as it isn't needed for any third party compilation yet.

**Comment by gxyd** [comment]
So, we've fixed all the tests in this PR now, including the MRE here: https://github.com/lfortran/lfortran/pull/5569#issuecomment-2517138029

**Comment by gxyd** [comment]
Seems like the last fix was causing some issues in our third party tests, so I've reverted it. I'll handle it in a separate issue, so that this is ready for a review now.

**Review by certik — APPROVED** [review]
I think this is good. The logic is getting quite complicated there, but I think it is correct and the overhead shouldn't be large here.

**Comment by gxyd** [comment]
> The logic is getting quite complicated there

Agreed, there a few corner cases, I should try to simplify the logic as well, once I've identified all the failing cases and some working solution to them (without overhead), after that I'll focus on simplifying the solution to it as well.
